runtime/strings/bytes.cpp

```diff
@@ -0,0 +1,176 @@
+#include <gmp.h>
+#include <cstdlib>
+#include <cstring>
+#include <cstdint>
+#include <algorithm>
+#include <stdexcept>
+
+extern "C" {
+
+#define KCHAR char
+
+  struct blockheader {
+    uint64_t len;
+  };
+
+  struct string {
+    blockheader b;
+    KCHAR data[0];
+  };
+
+  mpz_ptr move_int(mpz_t);
+
+  string *hook_BYTES_empty() {
+    static string empty;
+    empty.b.len = 0;
+    return &empty;
+  }
+
+  string *hook_STRING_concat(string *a, string *b);
+  mpz_ptr hook_STRING_length(string *a);
+  string *hook_STRING_substr(string *a, mpz_t start, mpz_t end);
+
+  uint32_t getTagForSymbolName(const char *);
+
+  uint64_t tag_big_endian() {
+    return (((uint64_t)getTagForSymbolName("LblbigEndianBytes{}")) << 32) | 1;
+  }
+
+  uint64_t tag_unsigned() {
+    return (((uint64_t)getTagForSymbolName("LblunsignedBytes{}")) << 32) | 1;
+  }
+
+  mpz_ptr hook_BYTES_bytes2int(string *b, uint64_t endianness, uint64_t signedness) {
+    mpz_t result;
+    mpz_init(result);
+    int order = endianness == tag_big_endian() ? 1 : -1;
+    mpz_import(result, b->b.len, order, 1, 0, 0, b->data);
+    if (signedness != tag_unsigned() && b->b.len != 0) {
+      bool msb;
+      if (endianness == tag_big_endian()) {
+        msb = b->data[0] & 0x80;
+      } else {
+        msb = b->data[b->b.len - 1] & 0x80;
+      }
+      if (msb) {
+        mpz_t max;
+        mpz_init_set_ui(max, 1);
+        mpz_mul_2exp(max, max, b->b.len*8);
+        mpz_sub(result, result, max);
+        mpz_clear(max);
+      }
+    }
+    return move_int(result);
+  }
+
+  string *hook_BYTES_int2bytes(mpz_t len, mpz_t i, uint64_t endianness) {
+    if (!mpz_fits_ulong_p(len)) {
+      throw std::invalid_argument("Integer overflow on len");
+    }
+    unsigned long len_long = mpz_get_ui(len);
```

Comment:
This common pattern of 
    
```
    if (!mpz_fits_ulong_p(len)) {
      throw std::invalid_argument("Integer overflow on len");
    }
    unsigned long len_long = mpz_get_ui(len);
```

Should be extracted out to a simple function like the one I had in strings, however, `gs` is a terrible name. Maybe `get_ui_checked`.

Reply:
lgtm